noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,309 @@
+        // We don't perform check for the overwritten value to be non-zero because the slots are siloed to the shield
+        // owner and hence the owner can interfere only with his own execution.
+        context.storage_write(slot, point);
+    }
+
+    #[aztec(public)]
+    fn send_to_shield(
+        from: AztecAddress,
+        token_id: Field,
+        owner_storage_slot_commitment: Field,
+        nonce: Field
+    ) {
+        // Perform authwit, owner and prepared shield checks
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let current_owner = storage.public_owners.at(token_id).read();
+        assert(current_owner.eq(from), "invalid NFT owner");
+
+        // Derive the shield slot from the owner storage slot commitment and the finalizer address (declared
+        // as `from` in this function)
+        let shield_slot = pedersen_hash(
+            [from.to_field(), owner_storage_slot_commitment],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+
+        // Read the hiding point from "transient" storage and check it's not empty to ensure the shield was prepared
+        let mut hiding_point: NFTNoteHidingPoint = context.storage_read(shield_slot);
+        assert(!is_empty(hiding_point), "shield not prepared");
+
+        // Set the original NFT owner to zero
+        storage.public_owners.at(token_id).write(AztecAddress::zero());
+
+        // Read the NFT hiding point from "transient" storage and add the token ID to it
+        hiding_point.add_token_id(token_id);
+
+        // Insert the finalized note
+        let note_hash = hiding_point.finalize();
+        context.push_note_hash(note_hash);
+
+        // At last we reset the shield to zero to achieve the effect of transient storage - kernels will squash the writes
+        context.storage_write(shield_slot, NFTNoteHidingPoint::empty());
+    }
+
+    /**
+     * Cancel a private authentication witness.
+     * @param inner_hash The inner hash of the authwit to cancel.
+     */
+    #[aztec(private)]
+    fn cancel_authwit(inner_hash: Field) {
+        let on_behalf_of = context.msg_sender();
+        let nullifier = compute_authwit_nullifier(on_behalf_of, inner_hash);
+        context.push_nullifier(nullifier);
+    }
+
+    #[aztec(private)]
+    fn transfer_from(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let nfts = storage.private_nfts;
+
+        let mut options = NoteGetterOptions::new();
+        options = options.select(NFTNote::properties().token_id, token_id, Option::none()).set_limit(1);
```

Comment:
This was much harder to read than it should've been. Created https://github.com/AztecProtocol/aztec-packages/issues/8413.